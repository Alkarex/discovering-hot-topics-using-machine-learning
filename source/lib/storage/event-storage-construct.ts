#!/usr/bin/env node
/**********************************************************************************************************************
 *  Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.                                           *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICNSE-2.0                                                                     *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import { Construct, Aws } from "@aws-cdk/core";
import { Bucket } from '@aws-cdk/aws-s3';
import { CfnDeliveryStream, CfnDeliveryStreamProps } from "@aws-cdk/aws-kinesisfirehose";
import { buildLambdaFunction, DefaultLogGroupProps } from '@aws-solutions-constructs/core';
import { Runtime, Code, Function } from "@aws-cdk/aws-lambda";
import { Role, ServicePrincipal, PolicyStatement, Policy, Effect } from "@aws-cdk/aws-iam";
import { LogGroup } from "@aws-cdk/aws-logs";
import { IDatabase } from "@aws-cdk/aws-glue";

export interface EventStorageProps {
    readonly compressionFormat: string,
    readonly prefix?: string,
    readonly processor?: boolean,
    readonly s3Bucket: Bucket,
    readonly keyArn?: string,

    //below props only required if data format is parquet
    /**
     * Convert data to Apache Parquet format
     */
     readonly convertData?: boolean,

    /**
     * The database name is required if Firehose should convert data to Apache Parquet
     */
    readonly database?: IDatabase,

    /**
     * Table is required to add dependency for Firehose creation
     */
    readonly tableName?: string
}

export class EventStorage extends Construct {

    private readonly _lambda: Function;
    private _firehose: CfnDeliveryStream;

    constructor (scope: Construct, id: string, props: EventStorageProps) {
        super(scope, id);

        // Setup Cloudwatch Log group & stream for Kinesis Firehose
        const cwLogGroup = new LogGroup(this, 'firehose-log-group', DefaultLogGroupProps());
        const cwLogStream = cwLogGroup.addStream('firehose-log-stream');

        // Setup the IAM Role for Kinesis Firehose
        const firehoseRole = new Role(this, 'FirehoseRole', {
            assumedBy: new ServicePrincipal('firehose.amazonaws.com'),
        });

        props.s3Bucket.grantReadWrite(firehoseRole); // add permissions to read/write to S3 bucket

        // Setup the IAM policy for Kinesis Firehose
        const firehosePolicy = new Policy(this, 'FirehosePolicy', {
            statements: [ new PolicyStatement({
                effect: Effect.ALLOW,
                actions: [
                    'logs:PutLogEvents'
                ],
                resources: [`arn:aws:logs:${Aws.REGION}:${Aws.ACCOUNT_ID}:log-group:${cwLogGroup.logGroupName}:log-stream:${cwLogStream.logStreamName}`]
            })]
        });

        // Attach policy to role
        firehosePolicy.attachToRole(firehoseRole);

        let firehoseGluePolicy: Policy;
        if (props.convertData && props.tableName !== undefined) {
            firehoseGluePolicy = new Policy(this, 'FirehoseGlueTablePolicy', {
                statements: [ new PolicyStatement({
                    effect: Effect.ALLOW,
                    actions: [ 'glue:GetTable', 'glue:GetTableVersion', 'glue:GetTableVersions' ],
                    resources: [
                        `arn:aws:glue:${Aws.REGION}:${Aws.ACCOUNT_ID}:table/${props.database!.databaseName}/${props.tableName}`,
                        `arn:aws:glue:${Aws.REGION}:${Aws.ACCOUNT_ID}:database/${props.database!.databaseName}`,
                        `arn:aws:glue:${Aws.REGION}:${Aws.ACCOUNT_ID}:catalog`
                    ]
                })]
            })

            firehoseGluePolicy.attachToRole(firehoseRole);
        }

        if (props.keyArn !== undefined) {
            const firehoseGlueKmsPolicy = new Policy(this, 'FirehoseGlueKms', {
                    statements: [ new PolicyStatement({
                    effect: Effect.ALLOW,
                    resources: [ props.keyArn ],
                    actions: [ 'kms:Decrypt' ]
                })]
            });

            firehoseGlueKmsPolicy.attachToRole(firehoseRole);
        }

        if (props?.processor) {
            this._lambda = buildLambdaFunction(this, {
                lambdaFunctionProps: {
                    runtime: Runtime.NODEJS_12_X,
                    handler: 'index.handler',
                    code: Code.fromAsset(`${__dirname}/../../lambda/storage-firehose-processor`),
                }
            });

            const lambdaProcessorPolicy = new Policy(this, 'LambdaProcessor', {
                statements: [new PolicyStatement({
                    effect: Effect.ALLOW,
                    resources: [ this._lambda.functionArn ],
                    actions: [ "lambda:InvokeFunction", "lambda:GetFunctionConfiguration"  ]
                })]
            });

            lambdaProcessorPolicy.attachToRole(firehoseRole);
        }

        // Setup the default Kinesis Firehose props
        const defaultKinesisFirehoseProps: CfnDeliveryStreamProps = {
            extendedS3DestinationConfiguration : {
                bucketArn: props.s3Bucket.bucketArn,
                bufferingHints: {
                    intervalInSeconds: 600,
                    sizeInMBs: 64
                },
                compressionFormat: props.compressionFormat,
                roleArn: firehoseRole.roleArn,
                cloudWatchLoggingOptions: {
                    enabled: true,
                    logGroupName: cwLogGroup.logGroupName,
                    logStreamName: cwLogStream.logStreamName
                },
                ...(props.prefix && {
                    prefix: `${props?.prefix}created_at=!{timestamp:yyyy-MM-dd}/`,
                    errorOutputPrefix: `${props.prefix}error/!{firehose:random-string}/!{firehose:error-output-type}/created_at=!{timestamp:yyyy-MM-dd}/`
                }),
                ...(props?.processor && {
                        processingConfiguration: {
                        enabled: true,
                        processors: [{
                            type: 'Lambda',
                            parameters: [{
                                parameterName: 'LambdaArn',
                                parameterValue: this._lambda.functionArn
                            }]
                        }]
                    }
                }),
                ...(props.convertData && {
                    dataFormatConversionConfiguration: {
                        inputFormatConfiguration: {
                            deserializer: {
                                openXJsonSerDe: {}
                            }
                        },
                        outputFormatConfiguration: {
                            serializer: {
                                parquetSerDe: {}
                            }
                        },
                        schemaConfiguration: {
                            databaseName: props.database!.databaseName,
                            tableName: props.tableName,
                            roleArn: firehoseRole.roleArn
                        }
                    }
                })
            }
        }


        // Override with the input props
        this._firehose = new CfnDeliveryStream(this, 'KinesisFirehose', defaultKinesisFirehoseProps);
        if (props.convertData && props.tableName !== undefined) {
            this._firehose.node.addDependency(firehoseGluePolicy!);
        }

    }

    public get kinesisFirehose(): CfnDeliveryStream {
        return this._firehose;
    }

    public get deliveryStreamName(): string {
        return this.kinesisFirehose.ref
    }

    public get deliveryStreamArn(): string {
        return this.kinesisFirehose.attrArn;
    }
}
