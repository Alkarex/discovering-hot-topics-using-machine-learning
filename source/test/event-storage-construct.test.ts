/**********************************************************************************************************************
 *  Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.                                           *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICNSE-2.0                                                                     *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import { SynthUtils } from '@aws-cdk/assert';
import { Stack } from '@aws-cdk/core';
import { EventStorage } from '../lib/storage/event-storage-construct';

import '@aws-cdk/assert/jest';

test ('Event Storage Construct with props', () => {
    const stack = new Stack();
    new EventStorage(stack, 'EventStorageTest', {
        compressionFormat: 'UNCOMPRESSED',
        prefix: 'test-prefix/'
    });
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});

test ('Event Storage Construct without props', () => {
    const stack = new Stack();
    const eventStorage = new EventStorage(stack, 'EventStorageTest', {
        compressionFormat: 'GZIP'
    });
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});

test ('Event Storage Construct with Lambda processor', () => {
    const stack = new Stack();
    const eventStorage = new EventStorage(stack, 'EventStorageTest', {
        compressionFormat: 'GZIP',
        processor: true,
        prefix: 'testPrefix/'
    });
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});