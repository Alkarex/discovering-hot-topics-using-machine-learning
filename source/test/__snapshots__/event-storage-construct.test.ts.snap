// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Event Storage Construct with Lambda processor 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters605f9425e66cabf7585f9cdfc6db214c0d9ed419cd53730f3e4595af862d35afArtifactHash70BCA2DB": Object {
      "Description": "Artifact hash for asset \\"605f9425e66cabf7585f9cdfc6db214c0d9ed419cd53730f3e4595af862d35af\\"",
      "Type": "String",
    },
    "AssetParameters605f9425e66cabf7585f9cdfc6db214c0d9ed419cd53730f3e4595af862d35afS3Bucket99E9F277": Object {
      "Description": "S3 bucket for asset \\"605f9425e66cabf7585f9cdfc6db214c0d9ed419cd53730f3e4595af862d35af\\"",
      "Type": "String",
    },
    "AssetParameters605f9425e66cabf7585f9cdfc6db214c0d9ed419cd53730f3e4595af862d35afS3VersionKey099DDC12": Object {
      "Description": "S3 key for asset version \\"605f9425e66cabf7585f9cdfc6db214c0d9ed419cd53730f3e4595af862d35af\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "EventStorageTestFirehosePolicy5A230078": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "logs:PutLogEvents",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:logs:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":log-group:",
                    Object {
                      "Ref": "EventStorageTestfirehoseloggroupACB75E85",
                    },
                    ":log-stream:",
                    Object {
                      "Ref": "EventStorageTestfirehoseloggroupfirehoselogstreamD3224E6C",
                    },
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "EventStorageTestFirehosePolicy5A230078",
        "Roles": Array [
          Object {
            "Ref": "EventStorageTestFirehoseRole9A5FD08D",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "EventStorageTestFirehoseRole9A5FD08D": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "firehose.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "EventStorageTestFirehoseRoleDefaultPolicyE272AE26": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject*",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "testBucketDF4D7D1A",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "testBucketDF4D7D1A",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "EventStorageTestFirehoseRoleDefaultPolicyE272AE26",
        "Roles": Array [
          Object {
            "Ref": "EventStorageTestFirehoseRole9A5FD08D",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "EventStorageTestKinesisFirehose442A2573": Object {
      "Properties": Object {
        "ExtendedS3DestinationConfiguration": Object {
          "BucketARN": Object {
            "Fn::GetAtt": Array [
              "testBucketDF4D7D1A",
              "Arn",
            ],
          },
          "BufferingHints": Object {
            "IntervalInSeconds": 300,
            "SizeInMBs": 64,
          },
          "CloudWatchLoggingOptions": Object {
            "Enabled": true,
            "LogGroupName": Object {
              "Ref": "EventStorageTestfirehoseloggroupACB75E85",
            },
            "LogStreamName": Object {
              "Ref": "EventStorageTestfirehoseloggroupfirehoselogstreamD3224E6C",
            },
          },
          "CompressionFormat": "GZIP",
          "ErrorOutputPrefix": "testPrefix/error/!{firehose:random-string}/!{firehose:error-output-type}/created_at=!{timestamp:yyyy-MM-dd}/",
          "Prefix": "testPrefix/created_at=!{timestamp:yyyy-MM-dd}/",
          "ProcessingConfiguration": Object {
            "Enabled": true,
            "Processors": Array [
              Object {
                "Parameters": Array [
                  Object {
                    "ParameterName": "LambdaArn",
                    "ParameterValue": Object {
                      "Fn::GetAtt": Array [
                        "EventStorageTestLambdaFunction0DDD544F",
                        "Arn",
                      ],
                    },
                  },
                ],
                "Type": "Lambda",
              },
            ],
          },
          "RoleARN": Object {
            "Fn::GetAtt": Array [
              "EventStorageTestFirehoseRole9A5FD08D",
              "Arn",
            ],
          },
        },
      },
      "Type": "AWS::KinesisFirehose::DeliveryStream",
    },
    "EventStorageTestLambdaFunction0DDD544F": Object {
      "DependsOn": Array [
        "EventStorageTestLambdaFunctionServiceRoleDefaultPolicy51A8DD20",
        "EventStorageTestLambdaFunctionServiceRole28BD7F58",
      ],
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W58",
              "reason": "Lambda functions has the required permission to write CloudWatch Logs. It uses custom policy instead of arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole with more tighter permissions.",
            },
          ],
        },
      },
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters605f9425e66cabf7585f9cdfc6db214c0d9ed419cd53730f3e4595af862d35afS3Bucket99E9F277",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters605f9425e66cabf7585f9cdfc6db214c0d9ed419cd53730f3e4595af862d35afS3VersionKey099DDC12",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters605f9425e66cabf7585f9cdfc6db214c0d9ed419cd53730f3e4595af862d35afS3VersionKey099DDC12",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "EventStorageTestLambdaFunctionServiceRole28BD7F58",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "TracingConfig": Object {
          "Mode": "Active",
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "EventStorageTestLambdaFunctionServiceRole28BD7F58": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "logs:CreateLogGroup",
                    "logs:CreateLogStream",
                    "logs:PutLogEvents",
                  ],
                  "Effect": "Allow",
                  "Resource": Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        "arn:aws:logs:",
                        Object {
                          "Ref": "AWS::Region",
                        },
                        ":",
                        Object {
                          "Ref": "AWS::AccountId",
                        },
                        ":log-group:/aws/lambda/*",
                      ],
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "LambdaFunctionServiceRolePolicy",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "EventStorageTestLambdaFunctionServiceRoleDefaultPolicy51A8DD20": Object {
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W12",
              "reason": "Lambda needs the following minimum required permissions to send trace data to X-Ray.",
            },
          ],
        },
      },
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "xray:PutTraceSegments",
                "xray:PutTelemetryRecords",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "EventStorageTestLambdaFunctionServiceRoleDefaultPolicy51A8DD20",
        "Roles": Array [
          Object {
            "Ref": "EventStorageTestLambdaFunctionServiceRole28BD7F58",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "EventStorageTestLambdaProcessor4069EDCC": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "lambda:InvokeFunction",
                "lambda:GetFunctionConfiguration",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "EventStorageTestLambdaFunction0DDD544F",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "EventStorageTestLambdaProcessor4069EDCC",
        "Roles": Array [
          Object {
            "Ref": "EventStorageTestFirehoseRole9A5FD08D",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "EventStorageTestfirehoseloggroupACB75E85": Object {
      "DeletionPolicy": "Retain",
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "EventStorageTestfirehoseloggroupfirehoselogstreamD3224E6C": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "LogGroupName": Object {
          "Ref": "EventStorageTestfirehoseloggroupACB75E85",
        },
      },
      "Type": "AWS::Logs::LogStream",
      "UpdateReplacePolicy": "Retain",
    },
    "testBucketDF4D7D1A": Object {
      "DeletionPolicy": "Retain",
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;

exports[`Event Storage Construct with props 1`] = `
Object {
  "Resources": Object {
    "EventStorageTestFirehosePolicy5A230078": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "logs:PutLogEvents",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:logs:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":log-group:",
                    Object {
                      "Ref": "EventStorageTestfirehoseloggroupACB75E85",
                    },
                    ":log-stream:",
                    Object {
                      "Ref": "EventStorageTestfirehoseloggroupfirehoselogstreamD3224E6C",
                    },
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "EventStorageTestFirehosePolicy5A230078",
        "Roles": Array [
          Object {
            "Ref": "EventStorageTestFirehoseRole9A5FD08D",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "EventStorageTestFirehoseRole9A5FD08D": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "firehose.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "EventStorageTestFirehoseRoleDefaultPolicyE272AE26": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject*",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "testBucketDF4D7D1A",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "testBucketDF4D7D1A",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "EventStorageTestFirehoseRoleDefaultPolicyE272AE26",
        "Roles": Array [
          Object {
            "Ref": "EventStorageTestFirehoseRole9A5FD08D",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "EventStorageTestKinesisFirehose442A2573": Object {
      "Properties": Object {
        "ExtendedS3DestinationConfiguration": Object {
          "BucketARN": Object {
            "Fn::GetAtt": Array [
              "testBucketDF4D7D1A",
              "Arn",
            ],
          },
          "BufferingHints": Object {
            "IntervalInSeconds": 300,
            "SizeInMBs": 64,
          },
          "CloudWatchLoggingOptions": Object {
            "Enabled": true,
            "LogGroupName": Object {
              "Ref": "EventStorageTestfirehoseloggroupACB75E85",
            },
            "LogStreamName": Object {
              "Ref": "EventStorageTestfirehoseloggroupfirehoselogstreamD3224E6C",
            },
          },
          "CompressionFormat": "UNCOMPRESSED",
          "ErrorOutputPrefix": "test-prefix/error/!{firehose:random-string}/!{firehose:error-output-type}/created_at=!{timestamp:yyyy-MM-dd}/",
          "Prefix": "test-prefix/created_at=!{timestamp:yyyy-MM-dd}/",
          "RoleARN": Object {
            "Fn::GetAtt": Array [
              "EventStorageTestFirehoseRole9A5FD08D",
              "Arn",
            ],
          },
        },
      },
      "Type": "AWS::KinesisFirehose::DeliveryStream",
    },
    "EventStorageTestfirehoseloggroupACB75E85": Object {
      "DeletionPolicy": "Retain",
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "EventStorageTestfirehoseloggroupfirehoselogstreamD3224E6C": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "LogGroupName": Object {
          "Ref": "EventStorageTestfirehoseloggroupACB75E85",
        },
      },
      "Type": "AWS::Logs::LogStream",
      "UpdateReplacePolicy": "Retain",
    },
    "testBucketDF4D7D1A": Object {
      "DeletionPolicy": "Retain",
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;

exports[`Event Storage Construct without props 1`] = `
Object {
  "Resources": Object {
    "EventStorageTestFirehosePolicy5A230078": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "logs:PutLogEvents",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:logs:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":log-group:",
                    Object {
                      "Ref": "EventStorageTestfirehoseloggroupACB75E85",
                    },
                    ":log-stream:",
                    Object {
                      "Ref": "EventStorageTestfirehoseloggroupfirehoselogstreamD3224E6C",
                    },
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "EventStorageTestFirehosePolicy5A230078",
        "Roles": Array [
          Object {
            "Ref": "EventStorageTestFirehoseRole9A5FD08D",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "EventStorageTestFirehoseRole9A5FD08D": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "firehose.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "EventStorageTestFirehoseRoleDefaultPolicyE272AE26": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject*",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "testBucketDF4D7D1A",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "testBucketDF4D7D1A",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "EventStorageTestFirehoseRoleDefaultPolicyE272AE26",
        "Roles": Array [
          Object {
            "Ref": "EventStorageTestFirehoseRole9A5FD08D",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "EventStorageTestKinesisFirehose442A2573": Object {
      "Properties": Object {
        "ExtendedS3DestinationConfiguration": Object {
          "BucketARN": Object {
            "Fn::GetAtt": Array [
              "testBucketDF4D7D1A",
              "Arn",
            ],
          },
          "BufferingHints": Object {
            "IntervalInSeconds": 300,
            "SizeInMBs": 64,
          },
          "CloudWatchLoggingOptions": Object {
            "Enabled": true,
            "LogGroupName": Object {
              "Ref": "EventStorageTestfirehoseloggroupACB75E85",
            },
            "LogStreamName": Object {
              "Ref": "EventStorageTestfirehoseloggroupfirehoselogstreamD3224E6C",
            },
          },
          "CompressionFormat": "GZIP",
          "RoleARN": Object {
            "Fn::GetAtt": Array [
              "EventStorageTestFirehoseRole9A5FD08D",
              "Arn",
            ],
          },
        },
      },
      "Type": "AWS::KinesisFirehose::DeliveryStream",
    },
    "EventStorageTestfirehoseloggroupACB75E85": Object {
      "DeletionPolicy": "Retain",
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "EventStorageTestfirehoseloggroupfirehoselogstreamD3224E6C": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "LogGroupName": Object {
          "Ref": "EventStorageTestfirehoseloggroupACB75E85",
        },
      },
      "Type": "AWS::Logs::LogStream",
      "UpdateReplacePolicy": "Retain",
    },
    "testBucketDF4D7D1A": Object {
      "DeletionPolicy": "Retain",
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;
